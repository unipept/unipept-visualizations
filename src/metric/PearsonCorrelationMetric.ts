import Metric from "./Metric";

export default class PearsonCorrelationMetric implements Metric {
    getDistance(matrix: number[][]): number[][] {
        let output: number[][] = [];

        for (let i = 0; i < matrix.length; i++) {
            let row: number[] = [];
            for (let j = 0; j <= i; j++) {
                row.push(this.getPearsonCorrelationBetween2Samples(matrix[i], matrix[j]));
            }
            output.push(row);
        }

        return output;
    }

    private getPearsonCorrelationBetween2Samples(x: number[], y: number[]) {
        let sum = (a: number, b: number) => a + b;
        let xMean = x.reduce(sum, 0) / x.length;
        let yMean = y.reduce(sum, 0) / y.length;

        let numerator = 0;
        let denominator = 0;

        for (let i = 0; i < x.length; i++) {
            numerator += (x[i] - xMean) * (y[i] - yMean);
            denominator += Math.sqrt(Math.pow((x[i] - xMean), 2)) * Math.sqrt(Math.pow((y[i] - yMean), 2));
        }

        return 1 - numerator / denominator;
    }
}
