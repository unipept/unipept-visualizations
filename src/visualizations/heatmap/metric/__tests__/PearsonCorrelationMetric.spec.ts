import TestDataGenerator from "./../../../../test/TestDataGenerator";
import PearsonCorrelationMetric from "./../PearsonCorrelationMetric";

describe("PearsonCorrelationMetric", () => {
    it("should produce lower triangular matrices", () => {
        const dataGenerator: TestDataGenerator = new TestDataGenerator();
        const data: number[][] = dataGenerator.generateSmall2DDataset();

        const pearsonMetric: PearsonCorrelationMetric = new PearsonCorrelationMetric();
        const distanceMatrix: number[][] = pearsonMetric.getDistance(data);

        for (let i: number = 0; i < distanceMatrix.length; i++) {
            expect(distanceMatrix[i].length).toBe(i + 1);
        }
    });

    it("should correctly calculate pearson correlation distances for small 2D dataset", () => {
        const dataGenerator: TestDataGenerator = new TestDataGenerator();
        const data: number[][] = dataGenerator.generateSmall2DDataset();

        const pearsonMetric: PearsonCorrelationMetric = new PearsonCorrelationMetric();
        const distanceMatrix: number[][] = pearsonMetric.getDistance(data);

        expect(distanceMatrix).toMatchSnapshot();
    });
});
