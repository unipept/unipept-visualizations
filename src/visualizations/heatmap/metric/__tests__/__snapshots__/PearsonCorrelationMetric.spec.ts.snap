// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PearsonCorrelationMetric > should correctly calculate pearson correlation distances for small 2D dataset 1`] = `
[
  [
    NaN,
  ],
  [
    NaN,
    0,
  ],
  [
    NaN,
    2,
    0,
  ],
  [
    NaN,
    2,
    0,
    0,
  ],
  [
    NaN,
    2,
    0,
    0,
    0,
  ],
]
`;
